src/providers/moduleProvider.ts

```diff
@@ -0,0 +1,161 @@
+import * as path from 'path';
+import * as vscode from 'vscode';
+import { ExecuteCommandParams, ExecuteCommandRequest } from 'vscode-languageclient';
+import { Utils } from 'vscode-uri';
+import { ClientHandler } from '../clientHandler';
+
+const LOCALMODULE = new vscode.ThemeIcon('symbol-folder', new vscode.ThemeColor('terminal.ansiBrightBlue'));
+const TFREGISTRY = new vscode.ThemeIcon('extensions-view-icon', new vscode.ThemeColor('terminal.ansiBrightMagenta'));
+const GITHUBMODULE = new vscode.ThemeIcon('github');
+
+class TerraformModule extends vscode.TreeItem {
+  constructor(
+    public label: string,
+    public provider: string,
+    public version: string,
+    public type: string,
+    public doclink: string,
+    public state: vscode.TreeItemCollapsibleState,
+    public readonly children?: [TerraformModule],
+  ) {
+    super(label, state);
+    if (this.version != '') {
+      this.tooltip = `${this.provider}@${this.version}`;
+      this.description = `${this.provider}@${this.version}`;
+    } else {
+      this.tooltip = `${this.provider}`;
+      this.description = `${this.provider}`;
+    }
+  }
+
+  iconPath = this.getIcon(this.type);
+
+  getIcon(type: string) {
+    const icon = this.terraformIcon();
+    switch (type) {
+      case 'tfregistry':
+        return icon;
+      case 'local':
+        return LOCALMODULE;
+      case 'github':
+        return GITHUBMODULE;
+      default:
+        return TFREGISTRY;
+    }
+  }
+
+  private terraformIcon() {
+    // need current extension path to find icon svg
+    // could possibly make this a custom icon
+    const myExtDir = vscode.extensions.getExtension('hashicorp.terraform').extensionPath;
+    const svg = vscode.Uri.file(path.join(myExtDir, 'assets', 'icons', 'terraform.svg'));
+    const icon = {
+      light: svg,
+      dark: svg,
+    };
+    return icon;
+  }
+}
+
+export class TerraformModuleProvider implements vscode.TreeDataProvider<TerraformModule> {
+  private _onDidChangeTreeData: vscode.EventEmitter<TerraformModule | undefined | null | void> =
+    new vscode.EventEmitter<TerraformModule | undefined | null | void>();
+  readonly onDidChangeTreeData: vscode.Event<TerraformModule | undefined | null | void> =
+    this._onDidChangeTreeData.event;
+
+  constructor(ctx: vscode.ExtensionContext, public handler: ClientHandler) {
+    ctx.subscriptions.push(
+      vscode.commands.registerCommand('terraform.modules.refreshList', () => this.refresh()),
+      vscode.commands.registerCommand('terraform.modules.documentation', (module: TerraformModule) => {
```

Comment:
```suggestion
      vscode.commands.registerCommand('terraform.modules.openDocumentation', (module: TerraformModule) => {
```
nitpick - I think it would be great if the name implied that a link will be opened